generative/maisi/scripts/utils.py

```diff
@@ -701,3 +702,45 @@ def load_autoencoder_ckpt(load_autoencoder_path):
             new_state_dict[k] = v
     checkpoint_autoencoder = new_state_dict
     return checkpoint_autoencoder
+
+
+def KL_loss(z_mu, z_sigma):
+    """
+    Compute the Kullback-Leibler (KL) divergence loss for a variational autoencoder (VAE).
+
+    The KL divergence measures how one probability distribution diverges from a second, expected probability distribution.
+    In the context of VAEs, this loss term ensures that the learned latent space distribution is close to a standard normal distribution.
+
+    Args:
+        z_mu (torch.Tensor): Mean of the latent variable distribution, shape [N,C,H,W,D] or [N,C,H,W].
+        z_sigma (torch.Tensor): Standard deviation of the latent variable distribution, same shape as 'z_mu'.
+
+    Returns:
+        torch.Tensor: The computed KL divergence loss, averaged over the batch.
+    """
+    kl_loss = 0.5 * torch.sum(
+        z_mu.pow(2) + z_sigma.pow(2) - torch.log(z_sigma.pow(2)) - 1,
```

Comment:
Changing `torch.log(z_sigma.pow(2))` to `torch.log(z_sigma.pow(2) + 1e-10)` may help to prevent numerical instability.

Reply:
Done, thanks